### PR TITLE
Add weida raw plugin

### PR DESCRIPTION
**Pull request by chenhn123**
Type of pull request:

- [x] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [ ] Feature
- [ ] Documentation

# Checklist 

- [x] Fill out `README.md` with update protocol
- [x] Fill out `README.md` with any custom quirks and flags
- [x] Fill out `README.md` with the vendor ID security value
- [ ] Implement `FuFirmware->write()` and include at least one fuzzer testcase in `src/fuzzing/firmware` for any custom `FuFirmware` subclass
- [ ] CI run of the plugin for at least one target
- [ ] Document targets CI isn't currently run and the reasons (i.e. minimum versions needed or distribution limitations etc).

### CODE REVIEW

File: plugins/weida-raw/fu-weida-raw-device.c [hunk 1]
```diff
@@ -0,0 +1,1093 @@
...
+#define FU_WEIDA_RAW_FOURCC_ID_FERA 0x41524546
+#define FU_WEIDA_RAW_FOURCC_ID_FBIN 0x4e494246
+#define FU_WEIDA_RAW_FOURCC_ID_FRMT 0x544D5246
+#define FU_WEIDA_RAW_FOURCC_ID_FRWR 0x52575246
+#define FU_WEIDA_RAW_FOURCC_ID_CNFG 0x47464E43
+
+typedef struct {
+	guint32 FourCC;
+	guint32 FileSize;
+	guint32 DataType;
```
Comment (hughsie) [inline plugins/weida-raw/fu-weida-raw-device.c:41]: struct fields need to be lower case please

File: plugins/weida-raw/fu-weida-raw-device.c [hunk 2]
```diff
@@ -0,0 +1,1093 @@
...
+		polling_timeout_ms = timeout_ms;
+	if (intv_ms)
+		polling_intv_ms = intv_ms;
+
+	do {
+		if (fu_weida_raw_w8760_set_device_mode(pdev, mode, error)) {
+			if (fu_weida_raw_w8760_get_device_mode(pdev, error) == mode)
+				return TRUE;
+		}
+
```
Comment (hughsie) [inline plugins/weida-raw/fu-weida-raw-device.c:360]: can this use fu_device_retry() please

File: plugins/weida-raw/fu-weida-raw-device.c [hunk 3]
```diff
@@ -0,0 +1,1093 @@
...
+	cmd[2] = (guint8)(address >> 12);
+	cmd[3] = (guint8)(((address & 0x0FFF) + size + 4095) >> 12);
+
+	if (!fu_weida_raw_device_set_feature(self, cmd, sizeof(cmd), error)) {
+		return FALSE;
+	} else {
+		if (!fu_weida_raw_w8760_wait_cmd_end(self, 0, 0, error)) {
+			return FALSE;
+		}
+	}
```
Comment (hughsie) [inline plugins/weida-raw/fu-weida-raw-device.c:419]: you don't need the }else{

File: plugins/weida-raw/fu-weida-raw-device.c [hunk 4]
```diff
@@ -0,0 +1,1093 @@
...
+	return TRUE;
+}
+
+static gboolean
+fu_weida_raw_w8760_set_flash_address(FuWeidaRawDevice *self, guint32 address, GError **error)
+{
+	guint8 cmd[10] = {0};
+	memset(&cmd, 0, sizeof(cmd));
```
Comment (hughsie) [inline plugins/weida-raw/fu-weida-raw-device.c:427]: you don;t need the memset if you used {0}

File: plugins/weida-raw/weida-raw.quirk [hunk 5]
```diff
@@ -0,0 +1,4 @@
+# Raw
+[HIDRAW\VEN_2575]
+Plugin = weida_raw
+GType = FuWeidaRawDevice
```
Comment (hughsie) [inline plugins/weida-raw/weida-raw.quirk:4]: I don't think you need to set the GType if there's only one to choose from.

File: plugins/weida-raw/fu-weida-raw-device.c [hunk 6]
```diff
@@ -0,0 +1,1093 @@
...
+static gboolean
+fu_weida_raw_device_ensure_status(FuWeidaRawDevice *self, GError **error)
+{
+	guint8 buf[] = {[0] = FU_WEIDA_RAW_REQ_DEV_INFO, [1 ... 63] = 0xff};
+
+	if (!fu_weida_raw_device_get_feature(self, buf, sizeof(buf), error))
+		return FALSE;
+	self->firmware_id = fu_memread_uint32(&buf[1], G_LITTLE_ENDIAN);
+	self->hardware_id = fu_memread_uint32(&buf[5], G_LITTLE_ENDIAN);
+	self->serial_number = fu_memread_uint32(&buf[9], G_LITTLE_ENDIAN);
```
Comment (hughsie) [inline plugins/weida-raw/fu-weida-raw-device.c:201]: fu_device_set_serial_number() here (with a string)

File: plugins/weida-raw/fu-weida-raw-common.h [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+fu_weida_raw_block_is_empty(const guint8 *data, gsize datasz);
+
+guint16
+fu_weida_raw_misr_16b(guint16 current_value, guint16 new_value);
+
+guint16
+fu_weida_raw_misr_for_bytes(guint16 current_value, guint8 *bytes, guint start, guint size);
+
+guint16
+fu_weida_raw_misr_for_halfwords(guint16 current_value, guint8 *buf, guint start, guint hword_count);
```
Comment (hughsie) [inline plugins/weida-raw/fu-weida-raw-common.h:24]: Can you use fu_misr16() instead of this please

File: plugins/weida-raw/fu-weida-raw-device.c [hunk 8]
```diff
@@ -0,0 +1,1093 @@
...
+#define FU_WEIDA_RAW_FOURCC_ID_RIFF 0x46464952
+#define FU_WEIDA_RAW_FOURCC_ID_WIF2 0x32464957
+#define FU_WEIDA_RAW_FOURCC_ID_WHIF 0x46494857
+#define FU_WEIDA_RAW_FOURCC_ID_INFO 0x4f464e49
+#define FU_WEIDA_RAW_FOURCC_ID_FSUM 0x4d555346
+#define FU_WEIDA_RAW_FOURCC_ID_FERA 0x41524546
+#define FU_WEIDA_RAW_FOURCC_ID_FBIN 0x4e494246
+#define FU_WEIDA_RAW_FOURCC_ID_FRMT 0x544D5246
+#define FU_WEIDA_RAW_FOURCC_ID_FRWR 0x52575246
+#define FU_WEIDA_RAW_FOURCC_ID_CNFG 0x47464E43
```
Comment (hughsie) [inline plugins/weida-raw/fu-weida-raw-device.c:36]: yup, move these to the .rs file too pls.

File: plugins/weida-raw/fu-weida-raw-device.c [hunk 9]
```diff
@@ -0,0 +1,1093 @@
...
+		return FALSE;
+	fu_dump_raw(G_LOG_DOMAIN, "GetFeature", data, datasz);
+
+	return TRUE;
+}
+
+static void
+fu_weida_raw_device_to_string(FuDevice *device, guint idt, GString *str)
+{
+	return;
```
Comment (hughsie) [inline plugins/weida-raw/fu-weida-raw-device.c:135]: you need to dump:

	gint32 dev_type;
	gint32 firmware_id;
	gint32 hardware_id;
	guint8 firmware_rev_ext;

here

File: plugins/weida-raw/fu-weida-raw-device.c (hunk 6 above)
Comment (hughsie) [reply plugins/weida-raw/fu-weida-raw-device.c:201]: ohh you do, ignore me. I think in this case don't store the serial_number in self, just read it locally to this function

File: plugins/weida-raw/fu-weida-raw-device.c [hunk 10]
```diff
@@ -0,0 +1,1093 @@
...
+	if ((fwid & 0xF0000000) == 0x40000000) {
+		fu_device_set_summary(FU_DEVICE(self), "CI4.0_TM4.0");
+		return FU_WEIDA_RAW_FW8760;
+	}
+	if ((fwid & 0xF0000000) == 0x50000000) {
+		fu_device_set_summary(FU_DEVICE(self), "CI5.0");
+		return FU_WEIDA_RAW_FW8790;
+	}
+
+	return FU_WEIDA_RAW_FW_NOT_SUPPORT;
```
Comment (hughsie) [inline plugins/weida-raw/fu-weida-raw-device.c:189]: I think this function should set `self->dev_type = ` and instead return true/false -- if false also setting a GError

File: plugins/weida-raw/fu-weida-raw-device.c (hunk 4 above)
Comment (chenhn123) [reply plugins/weida-raw/fu-weida-raw-device.c]: thanks solved

File: plugins/weida-raw/weida-raw.quirk (hunk 5 above)
Comment (chenhn123) [reply plugins/weida-raw/weida-raw.quirk]: solved

File: plugins/weida-raw/fu-weida-raw-device.c (hunk 3 above)
Comment (chenhn123) [reply plugins/weida-raw/fu-weida-raw-device.c]: solved

File: plugins/weida-raw/fu-weida-raw-device.c (hunk 1 above)
Comment (chenhn123) [reply plugins/weida-raw/fu-weida-raw-device.c]: Done

File: plugins/weida-raw/fu-weida-raw.rs [hunk 11]
```diff
@@ -19,3 +31,9 @@ enum FuWeidaRawCmd8760 {
     CALCULATE_FLASH_CHECKSUM = 0xD5,
 }
 
+#[repr(u16)]
+enum FuWeidaRawCmd8760u16 {
+    UNPROTECT_LOWER508K = 0x0044,
+    PROTECT_ALL = 0x007C,
```
Comment (hughsie) [inline plugins/weida-raw/fu-weida-raw.rs:37]: Weirdly, you specify the enums in CamelCase (which get converted to UPPER_CASE when we generate the header file) so these should be `UnprotectLower508k` and `ProtectAll`.

File: plugins/weida-raw/fu-weida-raw-device.c [hunk 12]
```diff
@@ -0,0 +1,1093 @@
...
+	FU_WDT8760_SET_DEVICE_MODE = 0xCA,
+	FU_WDT8760_REBOOT = 0xCE,
+	FU_WDT8760_SET_FLASH_ADDRESS = 0xD0,
+	FU_WDT8760_READ_FLASH = 0xD1,
+	FU_WDT8760_ERASE_FLASH = 0xD2,
+	FU_WDT8760_WRITE_FLASH = 0xD3,
+	FU_WDT8760_PROTECT_FLASH = 0xD4,
+	FU_WDT8760_CALCULATE_FLASH_CHECKSUM = 0xD5,
+
+} FU_WDT8760Command;
```
[the remark this answers is not in the document]
Comment (chenhn123) [reply plugins/weida-raw/fu-weida-raw-device.c]: solved

File: plugins/weida-raw/fu-weida-raw-device.c (hunk 2 above)
Comment (chenhn123) [reply plugins/weida-raw/fu-weida-raw-device.c]: solved

File: plugins/weida-raw/fu-weida-raw.rs (hunk 11 above)
Comment (chenhn123) [reply plugins/weida-raw/fu-weida-raw.rs]: solved

File: plugins/weida-raw/fu-weida-raw-common.h (hunk 7 above)
Comment (chenhn123) [reply plugins/weida-raw/fu-weida-raw-common.h:24]: Hi, In my branch(my_new_change) we don't have fu_misr16() function
I think I need to sync and merge the project or cherry-pick from main branch,
But I have no relevant experience during the pull request.

Could you tell me how best to do this?

Comment (hughsie) [reply plugins/weida-raw/fu-weida-raw-common.h:24]: Sure! You can rebase -- e.g. something like:

    git remote add upstream
    git fetch upstream
    git rebase upstream/main

File: plugins/weida-raw/fu-weida-raw-device.c (hunk 10 above)
Comment (chenhn123) [reply plugins/weida-raw/fu-weida-raw-device.c]: Done

File: plugins/weida-raw/fu-weida-raw-common.h (hunk 7 above)
Comment (chenhn123) [reply plugins/weida-raw/fu-weida-raw-common.h]: thanks, soloved

File: plugins/weida-raw/fu-weida-raw-device.c (hunk 9 above)
Comment (chenhn123) [reply plugins/weida-raw/fu-weida-raw-device.c:107]: soloved

**Comment by hughsie** [comment]
@chenhn123 I've sent you an invite to the project, and will create a new `weida/raw` branch we can work on together.

**Comment by hughsie** [comment]
Lets carry on in https://github.com/fwupd/fwupd/pull/7275

**Pull request closed without merging**